IncomingWebSocket.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+
+const EventEmitter = require('events');
+const {
+  kConnectionOpened,
+  kMessageReceived,
+  kError,
+  kClientClosed,
+  kDrainMessage,
+  kDrainCompleted,
+  kUpstreamClosed
+} = require('./constants');
+const logger = require('./loggerFactory');
+const Queue = require('./Queue');
+
+/**
+  * Incoming connection to the proxy will be treated as an IncomingWebSocket.
+  * This will be the object having 1:1 relationship with the
+  * OutgoingWebSocket object. Since each client connection will
+  * have its own unique upstream.
+  *
+  * Each Incoming and Outgoing WebSocket is tied with a context. Context is an
+  * object which store the state for the session.
+  */
+class IncomingWebSocket extends EventEmitter {
+  constructor(socket, request) {
+    super();
+    this.socket = socket;
+    this.request = request;
+    this.queue = new Queue();
+    this.teardown = false;
+    this.registerListeners();
+  }
+
+  /**
+   * Registers the socket listeners.
+   */
+  registerListeners() {
+    this.socket.on('open', this.openHandler.bind(this));
+    this.socket.on('message', this.messageHandler.bind(this));
+    this.socket.on('close', this.closeHandler.bind(this));
+    this.socket.on('error', this.errorHandler.bind(this));
+  }
+
+  /**
+   * Triggers when socket connection is opened.
+   */
+  openHandler() {
+    this.emit(kConnectionOpened);
+  }
+
+  /**
+   * Triggers when message is received on socket.
+   *
+   * @param {string} msg
+   */
+  messageHandler(msg) {
+    this.emit(kMessageReceived, msg);
+  }
+
+  /**
+   * Triggers when socket connection is closed.
+   *
+   * @param {number} code
+   * @param {string} msg
+   */
+  closeHandler(code, msg) {
+    if (!this.teardown) {
+      this.emit(kClientClosed, code, msg);
+    }
+
+  }
+
+  /**
+   * Triggers when error occured on socket.
+   */
+  errorHandler() {
+    this.emit(kError);
+  }
+
+  /**
+   * Sets the incoming socket.
+   *
+   * @param {WebSocket} socket
+   * @param {object} request
+   */
+  setSocket(socket, request) {
+    this.socket = socket;
+    this.request = request;
+    this.registerListeners();
+  }
+
+  /**
+   * Adds message to queue.
+   *
+   * @param {string} msg
+   */
+  addToQueue(msg) {
+    this.queue.enqueue(msg);
+  }
+
+  /**
+   * Drains the queue and emits completed event.
+   */
+  drainQueue() {
+    while (!this.queue.isEmpty()) {
+      this.emit(kDrainMessage, this.queue.dequeue());
+    }
+    this.emit(kDrainCompleted);
+  }
+
+  /**
+   * Closes the socket connection.
+   */
+  close() {
+    this.teardown = true;
+    this.socket.close(1001, '');
```

Comment:
Check if this explicit close code needs to be sent or can be avoided?

Reply:
Close not required, only terminate is required. Removing close

Reply:
removed